services/backend/src/recommendation_system/ml_models/alpha/two_tower.py

```diff
@@ -36,54 +54,543 @@ def forward_user(self, user_tensor):
+                        torch.ones_like(
+                            engagement_type_vector
+                        ),  # engaged 500ms => 2.5s
+                        torch.zeros_like(engagement_type_vector),  # bad engagement
+                    ),
+                ),  # millisecond engaged with
+            ),
+        )
+
+    def forward(
+        self,
+        user_embedding,
+        content_embedding,
+        targets,
+        pos_weight=None,
+        neg_weight=None,
+        with_debug=False,
+    ):
+        noise_factor = 0.0005
+        user_embedding += noise_factor * torch.randn(*user_embedding.shape)
+        content_embedding += noise_factor * torch.randn(*content_embedding.shape)
+
+        cosine_sim = F.cosine_similarity(user_embedding, content_embedding, dim=1)
+
+        # Contrastive loss
+        pos_loss = targets * torch.pow(
+            torch.clamp(self.margin - cosine_sim, min=0.0), 2
+        )
+        neg_loss = (1 - targets) * torch.pow(cosine_sim, 2)
+
+        # If weights are provided, apply them to the positive and negative losses
+        if pos_weight is not None:
+            pos_loss *= pos_weight
+        if neg_weight is not None:
+            neg_loss *= neg_weight
+
+        ## Negative sampling
+        neg_samples_indices = torch.randint(
+            0, len(content_embedding), (user_embedding.size(0), self.num_neg_samples)
+        )
+        neg_samples = content_embedding[neg_samples_indices, :]
+        neg_cosine_sim = F.cosine_similarity(
+            user_embedding.unsqueeze(1), neg_samples, dim=2
+        )
+        neg_loss += torch.pow(neg_cosine_sim, 2).sum(dim=1)
+
+        loss_contrastive = torch.mean(pos_loss + neg_loss)
+
+        # Regularization terms
+        reg_user = torch.norm(user_embedding, p=2)
+        reg_content = torch.norm(content_embedding, p=2)
+        regularization_loss = reg_user + reg_content
+
+        # orthognal loss of content
+        ortho_reg = torch.norm(
+            torch.mm(content_embedding, content_embedding.t())
+            - torch.eye(content_embedding.size(0))
+        )
+
+        total_loss = (
+            loss_contrastive
+            + self.lambda_reg * regularization_loss
+            + self.lambda_orthog * ortho_reg
+        )
+        if with_debug:
+            print(
+                f"""losses are:
+              {loss_contrastive},
+              {self.lambda_reg * regularization_loss},
+              {self.lambda_orthog * ortho_reg},
+          """
+            )
+
+        return total_loss
+
+
 # Model Wrapper
 class ModelWrapper:
     def __init__(self, model_path=""):
         if not model_path:
             self.model = DummyTwoTowerModel()
         else:
             self.model = TwoTowerModel()
-            self.model.load_state_dict(torch.load(model_path, map_location=torch.device('cpu')))
+            self.model.load_state_dict(
+                torch.load(model_path, map_location=torch.device("cpu"))
```

Comment:
where is your model_path? where is your model, I don't see it uploaded here

Reply:
Thank you for highlighting this Professor, so that I may add the correct path - Should I specify the full path as "\src\recommendation_system\ml_models\alpha\two_tower_trained_saved.pth" in the ModelWrapper constructor?

Reply:
See other PRs from other teams